src/time_driver.rs

```diff
@@ -0,0 +1,246 @@
+            if pending.tfpnd().bit_is_set() {
+                self.trigger_alarm(cs);
+            }
+        });
+    }
+
+    fn next_period(&self, cs: CriticalSection) {
+        // We only modify the period from the timer interrupt, so we know this can't race.
+        let period = self.period.load(Ordering::Relaxed) + 1;
+        self.period.store(period, Ordering::Relaxed);
+        let t = (period as u64) << 15;
+
+        let alarm = self.alarm.borrow(cs);
+        let at = alarm.timestamp.get();
+
+        if at < t + 0xc000 {
+            // Enable alarm interrupt, the compare-match value has already been set earlier.
+            regs().tn_ien().modify(|_, w| w.tfien().set_bit());
+        }
+    }
+
+    fn trigger_alarm(&self, cs: CriticalSection) {
+        let mut next = self.queue.borrow(cs).borrow_mut().next_expiration(self.now());
+        while !self.set_alarm(cs, next) {
+            next = self.queue.borrow(cs).borrow_mut().next_expiration(self.now());
+        }
+    }
+
+    fn set_alarm(&self, cs: CriticalSection, timestamp: u64) -> bool {
+        let r = regs();
+        self.alarm.borrow(cs).timestamp.set(timestamp);
+        let t = self.now();
+
+        if timestamp <= t {
+            // If alarm timestamp has passed the alarm will not fire.
+            // Disarm the alarm and return `false` to indicate that.
+            regs().tn_ien().modify(|_, w| w.tfien().clear_bit());
+
+            self.alarm.borrow(cs).timestamp.set(u64::MAX);
+
+            return false;
+        }
+
+        // Write the alarm compare value regardless of whether we're going to enable it now or not.
+        // This way, when we enable it later, the right value is already set.
+        // Cast it to u16 grabbing the lsb's, and subtract because our counter counts down.
+        let counter = 0xffff - (timestamp as u16);
+        r.tn_cpb().write(|w| unsafe { w.bits(counter) });
+
+        // Enable it if it'll happen soon. Otherwise, `next_period` will enable it.
+        let diff = timestamp - t;
+        regs().tn_ien().modify(|_, w| w.tfien().bit(diff < 0xc000));
+
+        // Reevaluate if the alarm timestamp is still in the future
+        let t = self.now();
+        if timestamp <= t {
+            // If alarm timestamp has passed since we set it, we have a race condition and
+            // the alarm may or may not have fired.
+            // Disarm the alarm and return `false` to indicate that.
+            // It is the caller's responsibility to handle this ambiguity.
+            regs().tn_ien().modify(|_, w| w.tfien().clear_bit());
+
+            self.alarm.borrow(cs).timestamp.set(u64::MAX);
+
+            return false;
+        }
+
+        // We're confident the alarm will ring in the future.
+        true
+    }
+}
+
+impl Driver for MultiFunctionTimerDriver {
+    fn now(&self) -> u64 {
+        let period = self.period.load(Ordering::Relaxed);
+        compiler_fence(Ordering::Acquire);
```

Comment:
Another copy? This can be a load with the `Acquire` ordering. Not sure if that actually changes anything

Reply:
I believe this Acquire relates to the volatile read of the counter, and not to that of the period.